server/sonar-server/src/main/java/org/sonar/server/email/ws/UpdateAction.java

```diff
@@ -0,0 +1,139 @@
+
+  private final DbClient dbClient;
+  private final UserSession userSession;
+
+  // TODO remove this when EmailSettings won't use Settings in memory to get settings
+  private final Settings settings;
+
+  public UpdateAction(DbClient dbClient, UserSession userSession, Settings settings) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+    this.settings = settings;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction("update")
+      .setDescription("Update email settings<br>" +
+        "Requires 'Administer System' permission")
+      .setSince("6.1")
+      .setInternal(true)
+      .setPost(true)
+      .setHandler(this);
+
+    action.createParam(PARAM_HOST)
+      .setDescription("SMTP host. Leave blank to disable email sending.")
+      .setExampleValue("smtp.gmail.com");
+
+    action.createParam(PARAM_PORT)
+      .setDescription("Port number to connect with SMTP server.")
+      .setExampleValue(SMTP_PORT_DEFAULT);
+
+    action.createParam(PARAM_SECURE)
+      .setDescription("Whether to use secure connection and its type.")
+      .setPossibleValues(SSL_VALUE, STARTTLS_VALUE)
+      .setExampleValue(SSL_VALUE);
+
+    action.createParam(PARAM_USERNAME)
+      .setDescription("Username to use with authenticated SMTP.")
+      .setExampleValue("my_username");
+
+    action.createParam(PARAM_PASSWORD)
+      .setDescription("Username to use with authenticated SMTP.")
+      .setExampleValue("my_password");
+
+    action.createParam(PARAM_FROM)
+      .setDescription("Emails will come from this address. For example - \"noreply@sonarsource.com\". Note that server may ignore this setting (like does GMail).")
+      .setExampleValue(FROM_DEFAULT);
+
+    action.createParam(PARAM_PREFIX)
+      .setDescription("Prefix will be prepended to all outgoing email subjects.")
+      .setExampleValue(PREFIX_DEFAULT);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkPermission(GlobalPermissions.SYSTEM_ADMIN);
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      save(dbSession, SMTP_HOST, request.param(PARAM_HOST));
+      save(dbSession, SMTP_PORT, request.hasParam(PARAM_PORT) ? Integer.toString(request.paramAsInt(PARAM_PORT)) : null);
+      save(dbSession, SMTP_SECURE_CONNECTION, request.param(PARAM_SECURE));
+      save(dbSession, SMTP_USERNAME, request.param(PARAM_USERNAME));
+      save(dbSession, SMTP_PASSWORD, request.param(PARAM_PASSWORD));
+      save(dbSession, FROM, request.param(PARAM_FROM));
+      save(dbSession, PREFIX, request.param(PARAM_PREFIX));
+      dbSession.commit();
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+    response.noContent();
+  }
+
+  private void save(DbSession dbSession, String key, @Nullable String value) {
+    settings.setProperty(key, value);
+    if (value != null) {
```

Comment:
I should know this, but… Don't we want to make the difference between the parameter that is not sent, and the parameter with an empty value? Then value should not be `@Nullable`